artemis/modules/domain_expiration_scanner.py

```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env python3
+import datetime
+from typing import Any, Dict
+
+from karton.core import Task
+from whois import query  # type: ignore
+
+from artemis.binds import TaskStatus, TaskType
+from artemis.config import Config
+from artemis.domains import is_main_domain
+from artemis.module_base import ArtemisBase
+
+
+class DomainExpirationScanner(ArtemisBase):
+    """
+    Alerts if domain expiration date is coming.
+    """
+
+    identity = "domain_expiration_scanner"
+    filters = [{"type": TaskType.DOMAIN.value}]
+
+    def run(self, current_task: Task) -> None:
+        domain = current_task.get_payload(TaskType.DOMAIN)
+        if is_main_domain(domain):
+            now = datetime.datetime.now()
+            result: Dict[str, Any] = {}
+            domain_data = query(domain)
```

Comment:
Did you check:

- what are the rate limits for whois? 
- do they differ between registries?
- what pywhois does if limits are hit?

Reply:
- Whois limits are not oficially disclosed. 
- They may differ between registries - for .com there is no confirmed information, for .pl - https://www.dns.pl/en/whois/FAQ (from a single IP address - 100 per day)
- I assume that you're asking about library used here. It is python-whois, which is a wrapper around system's whois. This library does not add its own limits - it just passes on all requests to whois.

Reply:
If for .pl the limit is 100 per day, how can we handle the case of timeouts?

Reply:
We have two things to manipulate - 1. querying cache (available directly in python-whois) - it defaults to 48 h 2. waiting time - to reverse a ban - after using all requests, we get a particular exception, so we can wait a certain amount of time - in this case probably 24h + 30 min (they grant credits every ca.15 min so it's safe to add 30 min). But it seems that they count time after 00:00 as a day.